### PR TITLE
Refactor: Break Session Table HTML down into structured data

### PR DESCRIPTION
**Pull request by Gweaton**
## What does this pull request do?

Refactors SP intervention progress presenter to no longer pass HTML to the View layer, instead constructing the HTML inside the view layer from structured data passed by the presenter.

## What is the intent behind these changes?

Passing structured data around the application rather than long HTML strings feels safer and less brittle to test, as we're not testing whole strings and asserting that the HTML attributes are in a specific order. It should be the View's responsibility to add HTML to the data passed to it.

### CODE REVIEW

File: server/routes/serviceProviderReferrals/interventionProgressView.ts [hunk 1]
```diff
@@ -110,12 +110,24 @@ export default class InterventionProgressView {
...
+          { html: this.linkHtml(row.links) },
         ]
       }),
     }
   }
 
+  private linkHtml(links: { text: string; href: string }[]): string {
+    const linkArray = links.map(link => {
+      if (link.text && link.href) {
+        return `<a class="govuk-link" href="${link.href}">${link.text}</a>`
```
Comment (lawrence-forooghian) [inline server/routes/serviceProviderReferrals/interventionProgressView.ts:122]: These both need a `ViewUtils.escape`.

**Review by lawrence-forooghian — CHANGES_REQUESTED** [review]
Love it, a good refactor.

File: server/routes/serviceProviderReferrals/interventionProgressView.ts [hunk 2]
```diff
@@ -110,12 +110,24 @@ export default class InterventionProgressView {
...
-          { html: `${row.linkHtml}` },
+          { html: this.linkHtml(row.links) },
         ]
       }),
     }
   }
 
+  private linkHtml(links: { text: string; href: string }[]): string {
+    const linkArray = links.map(link => {
+      if (link.text && link.href) {
```
Comment (lawrence-forooghian) [inline server/routes/serviceProviderReferrals/interventionProgressView.ts:121]: When would this condition be false?

Comment (Gweaton) [reply server/routes/serviceProviderReferrals/interventionProgressView.ts:121]: When link/href is an empty string (which we set it to if we don't want to display a link)

File: server/routes/serviceProviderReferrals/interventionProgressView.ts (hunk 1 above)
Comment (Gweaton) [reply server/routes/serviceProviderReferrals/interventionProgressView.ts:122]: Even if there's no user input? These are coming from the urls / text fields we've hardcoded in the presenter.

File: server/routes/serviceProviderReferrals/interventionProgressView.ts (hunk 2 above)
Comment (lawrence-forooghian) [reply server/routes/serviceProviderReferrals/interventionProgressView.ts:121]: Hmm, could we use `null` to indicate this instead? It's clearer, I think.

Comment (lawrence-forooghian) [reply server/routes/serviceProviderReferrals/interventionProgressView.ts:121]: (and it's what we do elsewhere)

File: server/routes/serviceProviderReferrals/interventionProgressView.ts (hunk 1 above)
Comment (lawrence-forooghian) [reply server/routes/serviceProviderReferrals/interventionProgressView.ts:122]: Yeah, what if there are non HTML-safe characters in the text, for example? They still need to be escaped.

Comment (Gweaton) [reply server/routes/serviceProviderReferrals/interventionProgressView.ts]: Ok, have updated this

File: server/routes/serviceProviderReferrals/interventionProgressView.ts (hunk 2 above)
Comment (Gweaton) [reply server/routes/serviceProviderReferrals/interventionProgressView.ts]: Sorry, I got this confused with the PP code path (where we do set them to empty strings), they're always set here. I've removed the check here and added a separate commit for making the PP ones nullable.

**Review by lawrence-forooghian — APPROVED** [review]

**Pull request merged**